Network/Socket/ReadShow.hs

```diff
@@ -19,77 +25,76 @@ eqFst x = \(x',_) -> x' == x
 eqSnd :: Eq b => b -> (a, b) -> Bool
 eqSnd y = \(_,y') -> y' == y
 
--- | Return RHS element that is paired with provided LHS,
---   or apply a default fallback function if the list is partial
-lookForward :: Eq a => (a -> b) -> [Pair a b] -> a -> b
-lookForward defFwd ps x
-  = case filter (eqFst x) ps of
-      (_,y):_ -> y
-      [] -> defFwd x
-
--- | Return LHS element that is paired with provided RHS,
---   or apply a default fallback function if the list is partial
-lookBackward :: Eq b => (b -> a) -> [Pair a b] -> b -> a
-lookBackward defBwd ps y
-  = case filter (eqSnd y) ps of
-      (x,_):_ -> x
-      [] -> defBwd y
-
-data Bijection a b
-   = Bijection
-     { defFwd :: a -> b
-     , defBwd :: b -> a
-     , pairs  :: [Pair a b]
-     }
-
--- | apply a bijection over an LHS-value
-forward :: (Eq a) => Bijection a b -> a -> b
-forward Bijection{..} = lookForward defFwd pairs
-
--- | apply a bijection over an RHS-value
-backward :: (Eq b) => Bijection a b -> b -> a
-backward Bijection{..} = lookBackward defBwd pairs
-
--- | show function for Int-like types that encodes negative numbers
--- with leading '_' instead of '-'
-_showInt :: (Show a, Num a, Ord a) => a -> String
-_showInt n | n < 0 = let ('-':s) = show n in '_':s
-           | otherwise = show n
-
--- | parse function for Int-like types that interprets leading '_'
---   as if it were '-' instead
-_readInt :: (Read a) => String -> a
-_readInt ('_':s) = read $ '-':s
-_readInt s = read s
-
-
--- | parse a quote-separated pair into a tuple of Int-like values
---   should not be used if either type might have
---   literal quote-characters in the Read pre-image
-_parse :: (Read a, Read b) => String -> (a, b)
-_parse xy =
-  let (xs, '\'':ys) = break (=='\'') xy
-   in (_readInt xs, _readInt ys)
-{-# INLINE _parse #-}
 
--- | inverse function to _parse
---   show a tuple of Int-like values as quote-separated strings
-_show :: (Show a, Num a, Ord a,  Show b, Num b, Ord b) => (a, b) -> String
-_show (x, y) = _showInt x ++ "'" ++ _showInt y
-{-# INLINE _show #-}
+-- | Unified automorphic involution over @Either a b@ that converts between
+--   LHS and RHS elements of a list of @Pair a b@ mappings and is the identity
+--   function if no matching pair is found
+lookBetween :: (Eq a, Eq b) => [Pair a b] -> Either a b -> Either a b
+lookBetween ps = \case
+  Left  x | (_,y):_ <- filter (eqFst x) ps -> Right y
+  Right y | (x,_):_ <- filter (eqSnd y) ps -> Left  x
+  z -> z
+
+type Bijection a b = [Pair a b]
 
-defShow :: Eq a => String -> (a -> b) -> (b -> String) -> (a -> String)
-defShow name unwrap sho = \x -> name ++ (sho . unwrap $ x)
+defShow :: Eq a => String -> (a -> b) -> (Int -> b -> ShowS) -> (Int -> a -> ShowS)
+defShow name unwrap shoPrec = \i x ->
+    showParen (i > app_prec) $ ((name ++ " ") ++) . (shoPrec (i+1) . unwrap $ x)
```

Comment:
Here, I'd like to suggest a minor style change, and a bug fix:
```suggestion
    showParen (i > app_prec) $ showString name . showChar ' ' . shoPrec (app_prec+1) (unwrap x)
```
Whatever the precedence of the context in which `Foo x` occurs, the `x` is in a constructor application context, and needs a precedence that's one higher than `app_prec`, which is often higher than `i+1`.
[ Note: `showString` is just a convenience alias for `(++)`, and `showChar` for `(:)`. ]